lib/statsd/instrument/compiled_metric.rb

```diff
@@ -0,0 +1,287 @@
+        # @return [String] The metric type character (e.g., "c" for counter)
+        def type
+          raise NotImplementedError, "Subclasses must implement #type"
+        end
+
+        # @return [Symbol] The method name to define (e.g., :increment)
+        def method_name
+          raise NotImplementedError, "Subclasses must implement #method_name"
+        end
+
+        # @return [Numeric, nil] The default value for the metric
+        def default_value
+          raise NotImplementedError, "Subclasses must implement #default_value"
+        end
+
+        # @return [Boolean] Whether the metric method accepts a block
+        def accepts_block?
+          false
+        end
+
+        # Defines the metric emission method - must be implemented by subclasses
+        # @param tags [Hash] The dynamic tags configuration
+        def define_metric_method(tags)
+          if tags.any?
+            define_dynamic_method(tags)
+          else
+            define_static_method
+          end
+        end
+
+        private
+
+        # Defines the metric method for metrics with dynamic tags
+        # Generates optimized code with tag caching
+        def define_dynamic_method(tags)
+          # Use the actual tag names as keyword arguments
+          tag_names = tags.keys
+          method = method_name
+          default_val = default_value
+          block_param = accepts_block? ? ", &block" : ""
+
+          method_code = <<~RUBY
+            def self.#{method}(#{tag_names.map { |name| "#{name}:" }.join(", ")}, value: #{default_val.inspect}, sample_rate: nil#{block_param})
+              # Compute hash of tag values for cache lookup
+              cache_key = #{tag_names.map { |name| "#{name}.hash" }.join(" ^ ")}
+
+              # Look up or create a PrecompiledDatagram
+              datagram =
+                if (cache = @tag_combination_cache)
+                  cached_datagram = cache[cache_key] ||=
+                    begin
+                      new_datagram = PrecompiledDatagram.new([#{tag_names.join(", ")}], @datagram_blueprint)
+
+                      # Clear cache if it grows too large to prevent memory bloat
+                      if cache.size > @max_cache_size
+                        StatsD.logger.warn("[CompiledMetric] Tag combination cache exceeded max size (\#{@max_cache_size}) for metric '\#{@name}', clearing cache")
+                        @tag_combination_cache = nil
+                      end
+
+                      new_datagram
+                    end
+
+                  # Hash collision detection
+                  if cached_datagram && #{tag_names.map.with_index { |name, i| "#{name} != cached_datagram.tag_values[#{i}]" }.join(" || ")}
+                    # Hash collision - fall back to creating a new datagram
+                    StatsD.logger.warn("[CompiledMetric] Hash collision detected for metric '\#{@name}' with cache_key \#{cache_key}")
+                    cached_datagram = nil
+                  end
```

Comment:
Here again, I think we would like a metric, this should happen quite rarely so having a metric would help keep an eye if this ever happens.